# Project: Profitable Apps for the Apple Store and Google Play

As a data analyst, I work for a company that build both Android and iOS mobile applications. Their main source of revenue comes from in-app adverts and they only build apps that are free to downlaod and install.

The goal of this project is to analyse the data to help our developers understand what type of applications are likely to attract more users.


## Opening and Exploring the Data

Below we have a function that simplifies the exploration of the datasets.

It takes in 4 parameters:
1. The dataset
2. The start and end which are both integers representing the indicies to slice the dataset
3. The rows and columns which have a default argument of False

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

We will now open and explore both datasets and have a look at the number of rows and columns. We also want to try and identify the columns that could assist us with our analysis.


In [4]:
opened_file1 = open("AppleStore.csv")
opened_file2 = open("googleplaystore.csv")
from csv import reader
read_file1 = reader(opened_file1)
read_file2 = reader(opened_file2)
appledata = list(read_file1)
googledata = list(read_file2)

In [5]:
explore_data(appledata, 0, 5, True)



['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7198
Number of columns: 16


We can see that the Apple store has 7198 apps with 16 columns. The `track_name`, `currency`, `price`, `rating_count_tot`, `rating_count_ver` and `prime_genre` could be benficial in our analysis.

In [6]:
explore_data(googledata, 0, 5, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10842
Number of columns: 13


We can see that the Google play store has 10842 apps with 13 columns. The `App`, `category`, `rating`, `Installs`, `Type`, `price`, `genre` could be benficial in our analysis.

## Deleting Wrong Data

The google play data set has a dedicated [discussion forum](https://www.kaggle.com/lava18/google-play-store-apps/discussion). In [one of the discussions](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) it has been flagged that there is missing data at row 10472. For me, this is row 10473 as I have not seperated the header from the rest of the rows. Let us print the header, the row with the error and a row with correct data to investigate.

In [7]:
print(googledata[0]) # header
print('\n')
print(googledata[10473]) # incorrect row
print('\n')
print(googledata[10474]) # correct row

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


The incorrect row refers to an app called 'Life Made WI-Fi Touchscreen Photo Frame'.

Upon review, we can highlight the following errors:

1. The `category` is missing from the row
2. `Genre` is blank data. It only has whitespace

Due to the data being incorrect, we will delete the row.

In [8]:
del googledata[10473] # run once to prevent deleting correct data

In [9]:
print(len(googledata))

10841


## Removing Duplicate Data

### Part One

When analysing data, there is the potential of there being duplicate data. This is the case with the Google Play dataset.

We will print a few duplicate rows to confirm this.

In [10]:
for app in googledata[1:]:
    name = app[0]
    if name == 'Instagram':
        print(app)
        print('\n')
    

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




Above we can see that the `Instagram` app has 4 duplicates.
We will now count the number of duplicates in the dataset.

In [11]:
duplicate_apps = []
unique_apps = []

for app in googledata[1:]:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps :', len(duplicate_apps))
print('\n')
print('Example of duplicate apps :', duplicate_apps[:15])
        

Number of duplicate apps : 1181


Example of duplicate apps : ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


The duplicate rows need to be removed. However, this will not be done randomly. If we refer to the duplicate instagram example we can see that fourth position of each row has different data. This row refers to the ratings row. With this information, we can assume that the data was collected at different times. We can use this as the criteria for removing duplicates. The more reviews the app has, the more recent the data should be.

### Part Two

Now, we want to remvove the duplicates. We will create a dictionary and use the application names as the unique keys. The dictionary value wil be the highest number of reviews for the app. We will then be able to create a new dataset which will have one entry per app.

In [12]:
reviews_max = {}

for app in googledata[1:]:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews      
    

The length of the google data set was 10841 rows(excluding the header row). We then deleted the row with missing data making the length of the data set 10840 rows.We had 1181 duplicate rows in the dataset. Therefore, we are expecting 9659 entries in the reviews_max dictionary.

Let us check to make sure this is the case.

In [13]:
print(len(reviews_max))

9659


Now we wil use the `reviews_max`dictionary we creeated above to remove the duplicate rows. Where we come across duplicates, we will keep the entries with the higher number of reviews.

We wil do the following:

1. create two empty lists, `android_clean` to store cleaned data and `already_added` to store app names.
2. loop through the android dataset. For every iteration:
    * assign app name to a variable named name
    * convert number of reviews to float and assisng it to a variable named `n_reviews`.
    * if the number of reviews is the same as the number of maximum reviews of the app name and name is not already in already_added we will append the row to `android_clean` and append the name to `already_added`.
    
This will help us keep track of apps already added.

In [14]:
android_clean = []
already_added = []

for app in googledata[1:]:
    name = app[0]
    n_reviews = float(app[3])
    
    if reviews_max[name] == n_reviews and name not in already_added:
        android_clean.append(app)
        already_added.append(name)

Let us now check if the expected number of rows, 9659, entered the clean dataset.

In [15]:
explore_data(android_clean, 0, 3, True)


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have 9659 rows of data as expected.

## Remove Non-English Apps

### Part One

Upon look at the datasets, you'll notice the names of some of the apps give the impression that they are not directed toward an English-speaking audience. 

Below, we see a couple of examples:

In [16]:
print(android_clean[4412][0])
print(android_clean[7940][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


We're not interested in keeping apps that are not english, so we'll remove them. One way to do this is to remove each app with a name containing a symbol that isn't commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).

All charcters that are specific to English text are encoded using the ASCII standard. Each ASCII charcter has a number coresponding between 0 - 127 asociated with it. We can use this to create a function to check if an app name has non-ASCII characters.

We will build this function below using the built-in ord() function to find out the corresponding encoding number of each character.

In [17]:
def check_name(string):
    for letter in string:
        if ord(letter) > 127:
            return False
    
    return True
        
print(check_name('Instagram'))
print(check_name('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_name('Docs To Go™ Free Office Suite'))
print(check_name('Instachat 😜'))

True
False
False
False


As can be seen above, the function is having some issues identifying some english apps, Docs To Go™ Free Office Suite and Instachat `😜` correctly due to emojies and characters like `™` as they fall outside the ASCII range.

We can see the fall outside the 127 ASCII range.

We will need to refactor our function.

### Part Two

In [18]:
def check_name(string):
    non_ascii = 0
    
    for letter in string:
        if ord(letter) > 127:
            non_ascii += 1
            
    if non_ascii > 3:
        return False
    else:
        return True
    
    
print(check_name('Docs To Go™ Free Office Suite'))
print(check_name('Instachat 😜'))
print(check_name('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


The function is still imperfect, but it is fairly effective for hat we need at this point in our analysis.

We will now use the check_name function to filter out non-english apps from both datasets.

In [19]:
google_english = []
apple_english = []

for app in android_clean:
    name = app[0]
    
    if check_name(name):
        google_english.append(app)
        
for app in appledata[1:]:
    name = app[1]
    
    if check_name(name):
        apple_english.append(app)
        
explore_data(google_english, 0, 4, True)
print('\n')
explore_data(apple_english, 0, 4, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '

We can see that we're left with 9614 Google apps and 6183 Apple apps.

## Isolating the Free Apps

As stated in the beginning, we only build apps that are free. The main source of revenue comes from in-app adds.
Therefore, we need to isolate the free apps from the dataset.

In [20]:
free_google_apps = []
free_apple_apps = []

for app in google_english:
    price = app[7]
    
    if price == '0':
        free_google_apps.append(app)


for app in apple_english:
    price = app[4]
    
    if price == '0.0':
        free_apple_apps.append(app)
        

print(len(free_google_apps))        
print(len(free_apple_apps))

8864
3222


There are 8864 apps remaining for Google and 3222 apps remaining for Apple. We are now ready to analyse the data.

## Most Common Apps by Genre

### Part One

As stated int he introduction, our aim is to determine the kinds of apps that are likely to attract more users due to teh fact that the number of users affects revenue.

In order to minimise risks and overhead, our validation strategy for app ideas have 3 steps:

1. Build a minimal android version of the app, and add it to the Google Play store.
2. If the app has a god response from users, we develop is further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Our end goal is to add the app on both Google Play and the App Store. Therefore, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let us begin the analysis by building looking at the most common genres for each market. In order to do this, we will need to build a frequncy table for the `prime_genre` column for the App Store and the `Genres` and `Category` column for The Google Play dataset.

### Part Two

We will create a function that takes two parameters: `dataset` (which will be a list of list) and `index` (Which will be an integer).

In [21]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
        
    table_percentages = {}
    
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
        
    return table_percentages
    

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Part Three

Now that our functions are created, we can have a look at the `prime_genre`, `Genres` & `Category` columns.

Lets start with the `prime_genre` column from the App Store


In [29]:
display_table(free_apple_apps, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The most common genre on the app store, as far as free english apps are concerned, is the Games app (58.16%) which accounts for more than half. The Entertainment apps is 7.88% followed by Photo and Video at 4.97%. Apps designed for Education only amount to 3.66% followed by Social Networking.

Based on the figures,the general impression is that we can see that most of the apps are made for entertainment purposes (games, photo and video, social networking, sports, music) whereas apps for practical purposes (education, shopping, utilities, productivity, lifestyle) are not in as much abundance.

However, the fact that there are more entertainment apps in the market than practical apps does not mean that there are more users for these applications.

Let's continue by examining the Genres and Category columns of the Google Play data set.

In [26]:
display_table(free_google_apps, 1) # Category column

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The situation is different on the Google Play store. There are more apps that are for practical purposes than entertainment at first glance. Upon further inspection, we can see the the Family category, which makes up 18.9% of apps, it mostly made of of kids games. Also, the family category has been merged with the games category.

![Image](https://play.google.com/store/apps/category/FAMILY?hl=en)

Practical apps seem to have a better representation on Google Play compared to the App Store. The frequency table for the `Genres` column further confirms this.

In [28]:
display_table(free_google_apps, 9) # Genre column

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the `Genres` and the `Category` columns is not very clear, but one thing we can notice is that the `Genres` column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.